File: Game/scripts/Title.lua
--- 对应 Title.cpp
-- require "Image"
-- require "ScenceBase"
require "GameDirector"
Title = {}
setmetatable(Title, {
    __call = function(self)
        ---@class Title:ScenceBase
        local obj = {}
        obj.mImage = Image(CMAKE_SOURCE_DIR .. "res/dummy.png")
        obj.mCursorPosistion = 1

        function obj:isA(what)
            return Title == what
        end

        ---comment
        ---@param director Director
        function obj:update(director)
            local next = self
            if Framework:isKeyTriggered(Enum.Keyboard.S) then
                self.mCursorPosistion = self.mCursorPosistion - 1
                if self.mCursorPosistion < 1 then
                    self.mCursorPosistion = 2
                end
            elseif Framework:isKeyTriggered(Enum.Keyboard.W) then
                self.mCursorPosistion = self.mCursorPosistion + 1
                if self.mCursorPosistion > 2 then
                    self.mCursorPosistion = 1
                end
            elseif Framework:isKeyTriggered(Enum.Keyboard.Space) then
                next = GameDirector(director:mode())
                if self.mCursorPosistion == 1 then
                    director:setMode(Enum.Mode.MODE_1P)
                elseif self.mCursorPosistion == 2 then
                    director:setMode(Enum.Mode.MODE_2P)
                else
                    error("no possible")
                end
            end
            self.mImage:draw()
            Framework:drawDebugString(1, 1, "[title] : boom man, press space for confirm")
            Framework:drawDebugString(2, 3, "one player");
            Framework:drawDebugString(2, 4, "two players");
            Framework:drawDebugString(1, self.mCursorPosistion + 2, ">");
            Framework:drawDebugString(20, 5, "Title.lua")
            return next
        end

        setmetatable(obj, {
            __index = ScenceBase(),
            __tostring = function() return "Title" end
        })
        return obj
    end
})

File: Example/Sequence/Title.cpp
#include "GameLib/GameLib.h"
#include "GameLib/Framework.h"
using namespace GameLib;

#include "Sequence/Title.h"
#include "Sequence/Parent.h"
#include "Image.h"
#include "Pad.h"
#include "SoundManager.h"

namespace Sequence{

Title::Title() : mImage( 0 ), mCursorPosition( 0 ){
	mImage = new Image( CMAKE_CURRENT_SOURCE_DIR "data/image/dummy.dds" );
}

Title::~Title(){
	SAFE_DELETE( mImage );
}

void Title::update( Parent* parent ){
	//获取输入
	if ( Pad::isTriggered( Pad::U ) ){
		--mCursorPosition;
		if ( mCursorPosition < 0 ){ //
 			mCursorPosition = 1;
		}
		//光标声音
		SoundManager::instance()->playSe( SoundManager::SE_CURSOR_MOVE );
	}else if ( Pad::isTriggered( Pad::D ) ){
		++mCursorPosition;
		if ( mCursorPosition > 1 ){ //1越过0就循环成0
			mCursorPosition = 0;
		}
		//光标声音
		SoundManager::instance()->playSe( SoundManager::SE_CURSOR_MOVE );
	}else if ( Pad::isTriggered( Pad::A ) ){
		parent->moveTo( Parent::NEXT_GAME );
		if ( mCursorPosition == 0 ){
			parent->setMode( Parent::MODE_1P );
		}else if ( mCursorPosition == 1 ){
			parent->setMode( Parent::MODE_2P );
		}else{
			HALT( "arienai" );
		}
		//决定音
		SoundManager::instance()->playSe( SoundManager::SE_SELECTION );
	}
	//绘制
	mImage->draw();
	//输出文字
	Framework f = Framework::instance();
	f.drawDebugString( 0, 0, "[title] : boom man" );//
	f.drawDebugString( 1, 2, "one player" );//
	f.drawDebugString( 1, 3, "two players" );//
	//写光标
	f.drawDebugString( 0, mCursorPosition + 2, ">" );
}

} //namespace Sequence